tensorzero-core/src/config/stored.rs

```diff
@@ -0,0 +1,135 @@
+//! Stored config types for backward-compatible deserialization of historical snapshots.
+//!
+//! When deprecating a config field:
+//! 1. Remove it from the `Uninitialized*` type (fresh configs will reject it)
+//! 2. Keep it in the `Stored*` type (snapshots can still load)
+//! 3. Implement migration in `From<Stored*> for Uninitialized*`
+//!
+//! The `From` implementations use explicit destructuring to ensure compile-time errors
+//! when fields are added or removed from either the stored or uninitialized types.
+
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use crate::config::gateway::UninitializedGatewayConfig;
+use crate::config::provider_types::ProviderTypesConfig;
+use crate::config::{
+    MetricConfig, PostgresConfig, UninitializedConfig, UninitializedFunctionConfig,
+    UninitializedToolConfig,
+};
+use crate::embeddings::UninitializedEmbeddingModelConfig;
+use crate::evaluations::UninitializedEvaluationConfig;
+use crate::inference::types::storage::StorageKind;
+use crate::model::UninitializedModelConfig;
+use crate::optimization::UninitializedOptimizerInfo;
+use crate::rate_limiting::UninitializedRateLimitingConfig;
+
+/// Top-level stored config type.
+///
+/// Only fields with deprecations in their subtree use `Stored*` types.
+/// Other fields re-use `Uninitialized*` types directly.
+#[derive(Debug, Default, Deserialize, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct StoredConfig {
```

Comment:
Consider this? https://github.com/tensorzero/tensorzero/pull/4848#discussion_r2560674896